### PR TITLE
Bind and BindAndHandle handles remote address in the same way now (#344). Refactoring to decrease code duplication.

### PR DESCRIPTION
**Pull request by rbudzko**
Additionally, I've "borrowed" @jrudolph tests.

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpAttributes.scala [hunk 1]
```diff
@@ -21,4 +21,6 @@ private[akka] object HttpAttributes {
   private[akka] def remoteAddress(address: Option[InetSocketAddress]) =
     Attributes(RemoteAddress(address))
 
+  private[akka] def empty =
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpAttributes.scala]: switch to `val`?

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 2]
```diff
@@ -57,7 +56,12 @@ class HttpExt(private val config: Config)(implicit val system: ActorSystem) exte
 
   private[this] final val DefaultPortForProtocol = -1 // any negative value
 
-  private def fuseServerLayer(settings: ServerSettings, connectionContext: ConnectionContext, log: LoggingAdapter)(implicit mat: Materializer): BidiFlow[HttpResponse, ByteString, ByteString, HttpRequest, NotUsed] = {
+  private[this]type ServerLayerBidiFlow = BidiFlow[HttpResponse, ByteString, ByteString, HttpRequest, NotUsed]
+  private[this]type ServerLayerFlow = Flow[ByteString, ByteString, Future[Done]]
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: space before "type" (2x). is the "[this]" needed here?

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpAttributes.scala (hunk 1 above)
Comment (viktorklang) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpAttributes.scala]: def?

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 3]
```diff
@@ -78,6 +82,36 @@ class HttpExt(private val config: Config)(implicit val system: ActorSystem) exte
...
+    .bind(
+      interface,
+      port,
+      settings.backlog,
+      settings.socketOptions,
+      halfClose = false,
+      settings.timeouts.idleTimeout
+    )
+
+  private def chosePort(port: Int, connectionContext: ConnectionContext) = if (port >= 0) port else connectionContext.defaultPort
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: "choosePort"

**Comment by jrudolph** [comment]
Very nice, indeed. Things get a bit crowded in Http.scala, so we might want to move some of the private methods somewhere else in the future.

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by rbudzko** [comment]
Fixed quirks pointed out.

**Comment by jrudolph** [comment]
LGTM. @2beaucoup can you resolve your review?

**Comment by 2beaucoup** [comment]
Sure. Done.

**Comment by jrudolph** [comment]
LGTM. Thanks a lot @rbudzko.

**Comment by ktoso** [comment]
Thanks a lot @rbudzko :)

**Pull request merged**